File: backend/src/models/Session.js
const mongoose = require('mongoose');

const sessionSchema = new mongoose.Schema({
    user_id: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    patient_id: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Patient',
        required: true
    },
    patientName: {
        type: String,
        required: true
    },
    session_title: {
        type: String,
        default: 'Medical Consultation'
    },
    session_summary: {
        type: String,
        default: null
    },
    status: {
        type: String,
        enum: ['recording', 'processing', 'completed', 'failed'],
        default: 'recording'
    },
    transcript_status: {
        type: String,
        enum: ['pending', 'processing', 'completed', 'failed'],
        default: 'pending'
    },
    transcript: {
        type: String,
        default: null
    },
    start_time: {
        type: Date,
        required: true
    },
    end_time: {
        type: Date,
        default: null
    },
    date: {
        type: String,
        default: function () {
            return this.start_time ? this.start_time.toISOString().split('T')[0] : new Date().toISOString().split('T')[0];
        }
    },
    duration: {
        type: String,
        default: null
    },
    templateId: {
        type: String,
        default: null
    },
    chunks: [{
        chunkNumber: Number,
        gcsPath: String,
        publicUrl: String,
        mimeType: String,
        uploadedAt: Date,
        size: Number
    }],
    totalChunks: {
        type: Number,
        default: 0
    },
    isComplete: {
        type: Boolean,
        default: false
    },
    clinical_notes: {
        type: Array,
        default: []
    },
    createdAt: {
        type: Date,
        default: Date.now
    },
    updatedAt: {
        type: Date,
        default: Date.now
    }
});

// Update the updatedAt timestamp before saving
sessionSchema.pre('save', function (next) {
    this.updatedAt = new Date();
    next();
});

module.exports = mongoose.model('Session', sessionSchema);

File: lib/models/session.dart
class RecordingSession {
  final String id;
  final String userId;
  final String patientId;
  final String patientName;
  final String status;
  final DateTime startTime;
  final DateTime? endTime;
  final String? duration;
  final String? templateId;
  final String? sessionTitle;
  final String? sessionSummary;
  final String? transcript;
  final String? transcriptStatus;
  final List<AudioChunkMetadata> chunks;
  final int totalChunks;
  final bool isComplete;

  RecordingSession({
    required this.id,
    required this.userId,
    required this.patientId,
    required this.patientName,
    required this.status,
    required this.startTime,
    this.endTime,
    this.duration,
    this.templateId,
    this.sessionTitle,
    this.sessionSummary,
    this.transcript,
    this.transcriptStatus,
    this.chunks = const [],
    this.totalChunks = 0,
    this.isComplete = false,
  });

  factory RecordingSession.fromJson(Map<String, dynamic> json) {
    return RecordingSession(
      id: json['id'] ?? json['_id'] ?? '',
      userId: json['user_id'] ?? '',
      patientId: json['patient_id'] ?? '',
      patientName: json['patientName'] ?? json['patient_name'] ?? '',
      status: json['status'] ?? 'recording',
      startTime: DateTime.parse(json['start_time'] ?? DateTime.now().toIso8601String()),
      endTime: json['end_time'] != null ? DateTime.parse(json['end_time']) : null,
      duration: json['duration'],
      templateId: json['templateId'],
      sessionTitle: json['session_title'],
      sessionSummary: json['session_summary'],
      transcript: json['transcript'],
      transcriptStatus: json['transcript_status'],
      chunks: (json['chunks'] as List<dynamic>?)
              ?.map((c) => AudioChunkMetadata.fromJson(c))
              .toList() ??
          [],
      totalChunks: json['totalChunks'] ?? 0,
      isComplete: json['isComplete'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'patient_id': patientId,
      'patientName': patientName,
      'status': status,
      'start_time': startTime.toIso8601String(),
      'end_time': endTime?.toIso8601String(),
      'duration': duration,
      'templateId': templateId,
      'session_title': sessionTitle,
      'session_summary': sessionSummary,
      'transcript': transcript,
      'transcript_status': transcriptStatus,
      'chunks': chunks.map((c) => c.toJson()).toList(),
      'totalChunks': totalChunks,
      'isComplete': isComplete,
    };
  }
}

class AudioChunkMetadata {
  final int chunkNumber;
  final String gcsPath;
  final String? publicUrl;
  final String mimeType;
  final DateTime uploadedAt;
  final int? size;

  AudioChunkMetadata({
    required this.chunkNumber,
    required this.gcsPath,
    this.publicUrl,
    required this.mimeType,
    required this.uploadedAt,
    this.size,
  });

  factory AudioChunkMetadata.fromJson(Map<String, dynamic> json) {
    return AudioChunkMetadata(
      chunkNumber: json['chunkNumber'] ?? 0,
      gcsPath: json['gcsPath'] ?? '',
      publicUrl: json['publicUrl'],
      mimeType: json['mimeType'] ?? 'audio/wav',
      uploadedAt: DateTime.parse(json['uploadedAt'] ?? DateTime.now().toIso8601String()),
      size: json['size'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'chunkNumber': chunkNumber,
      'gcsPath': gcsPath,
      'publicUrl': publicUrl,
      'mimeType': mimeType,
      'uploadedAt': uploadedAt.toIso8601String(),
      'size': size,
    };
  }
}
